skmorf/tree/_unsup_criterion.pyx

```diff
@@ -70,12 +69,22 @@ cdef class TwoMeans(Criterion):
             The dataset stored as a buffer for memory efficiency of shape
             (n_samples,).
         """
-        # need to implement
-        pass
-    
+        self.start = 0
+        self.pos = 0
+        self.end = 0
+
+        self.n_samples = 0
+        self.n_node_samples = 0
+        self.weighted_n_node_samples = 0.0
+        self.weighted_n_left = 0.0
+        self.weighted_n_right = 0.0
+
+        cdef SIZE_t k = 0
+        cdef SIZE_t max_n_classes = 0
+
     cdef int init(
         self,
-        const DOUBLE_t[:, ::1] X,
+        const DTYPE_t[:, :] X,
```

Comment:
```suggestion
        const DTYPE_t[:, ::1] X,
```

Reply:
Thank you @adam2392! I learned so much from this. One question I have is do you have to define `stride` to declare a variable as contiguous? In `_splitter.pyx`, there was a following class variable called X, then `init` would type it as an `object`. My guess was to keep the types the same everywhere, but judging from this usage of ambiguous typing for X, could we go back and forth from contiguous ([:, ::1]) and not contiguous ([:,:])?

```
cdef class BaseDenseSplitter(Splitter):
    cdef const DTYPE_t[:, :] X

    cdef SIZE_t n_total_samples

    cdef int init(
        self,
        object X,
```

Reply:
Oh good catch. Yeah we can leave it as is for now then. We can refactor if upstream sklearn changes it.